src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -2031,6 +2057,50 @@ ExtractDistributedInsertValuesRTE(Query *query)
 
 
 /*
+ * ExpandValuesLists expands VALUES lists by building new lists
+ * that include expressions from the target list for columns
+ * with a default expression.
+ */
+List *
+ExpandValuesLists(List *targetList, List *valuesLists)
+{
+	ListCell *valuesListCell = NULL;
+	List *expandedValuesLists = NIL;
+	ListCell *targetEntryCell = NULL;
+
+	foreach(valuesListCell, valuesLists)
+	{
+		List *valuesList = (List *) lfirst(valuesListCell);
+		List *expandedValuesList = NIL;
+
+		foreach(targetEntryCell, targetList)
+		{
+			TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+			Expr *targetExpr = targetEntry->expr;
+
+			if (IsA(targetExpr, Var))
+			{
+				/* expression from the VALUES section */
+				Var *targetListVar = (Var *) targetExpr;
+				targetExpr = list_nth(valuesList, targetListVar->varattno - 1);
```

Comment:
This loop's performance isn't going to be great, right? You're looping over the entire target list to process each row, and when you hit a var, you use `list_nth` (linear access) to find a node to copy.

Reply:
I kind of wonder if there isn't some way to rework this to operate in-place (injecting a cell into an extant list?) rather than rebuilding the lists. You're already destructively reassigning the `values_lists` field at the end of it all anyhow.